app/Http/Controllers/v1/User/ContributionController.php

```diff
@@ -0,0 +1,171 @@
+ *
+ * @copyright Copyright (c) 2018, SiMDE-UTC
+ * @license GNU GPL-3.0
+ */
+
+namespace App\Http\Controllers\v1\User;
+
+use App\Http\Controllers\v1\Controller;
+use Illuminate\Http\JsonResponse;
+use Illuminate\Http\Request;
+use App\Models\Semester;
+use App\Traits\Controller\v1\HasSemesters;
+use App\Traits\Controller\v1\HasUsers;
+use Illuminate\Support\Arr;
+
+class ContributionController extends Controller
+{
+    use HasUsers, HasSemesters;
+
+    /**
+     * Must be able to get user contributions.
+     */
+    public function __construct()
+    {
+        $this->middleware(
+            \Scopes::matchOneOfDeepestChildren('user-get-contributions', 'client-get-contributions'),
+            ['only' => ['all', 'get']]
+        );
+        $this->middleware(
+        \Scopes::matchOneOfDeepestChildren('user-create-contributions', 'client-create-contributions'),
+	        ['only' => ['create']]
+        );
+        $this->middleware(
+            \Scopes::matchOneOfDeepestChildren('user-edit-contributions', 'client-edit-contributions'),
+	        ['only' => ['edit']]
+        );
+        $this->middleware(
+            \Scopes::matchOneOfDeepestChildren('user-remove-contributions', 'client-remove-contributions'),
+        	['only' => ['remove']]
+        );
+    }
+
+    /**
+     * List contributions of the current user.
+     *
+     * @param Request $request
+     * @param string  $user_id
+     * @return JsonResponse
+     */
+    public function index(Request $request, string $user_id=null)
+    {
+        $user = $this->getUser($request, $user_id);
+        $ginger = \Ginger::userByEmail($user->email);
+        $contributions = [];
+        $semesters = Semester::get()
+            ->map(function($semester){
+                return $semester->hideData();
+            })->toArray();
+
+        foreach ($ginger->getContributions() as $contribution) {
+            $contributionSemesters = array_filter($semesters, function ($semester) use ($contribution) {
+                if (($semester['begin_at'] < $contribution->end_at) && ($semester['end_at'] > $contribution->begin_at)) {
+                    return true;
+                }
+            });
+
+            $newContribution = [
+                'start' => $contribution->begin_at,
+                'end' => $contribution->end_at,
+                'amount' => $contribution->money,
+                'semesters' => array_values($contributionSemesters),
+
+            ];
+
+            $contributions[] = $newContribution;
+        }
+
+        return response()->json($contributions);
+    }
+
+    /**
+     * Show a contribution depending on a given semester range and an optionnel user.
+     *
+     * @param Request $request
+     * @param string  $semester_id Work with id or name.
+     * @param string  $user_id
+     * @return JsonResponse
+     */
+    public function show(Request $request, string $semester_id="current", string $user_id=null)
```

Comment:
/!\ WRONG ORDER ! /!\

The `user_id` could be given by the client or not and after we give a `semester_id`.
If the `user_id` is `null`, you need to swap the values:
```php
        if (is_null($semester_id)) {
            list($user_id, $semester_id) = [$semester_id, $user_id];
        }
```

The `semester_id` must be supplied as it is a `show` method but could be `null` is `user_id` is not supplied so let `semester_id` `null` by default pls